### PR TITLE
planner/cascades: add appliedRuleSet in GroupExpr to avoid re-apply rules

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

This PR  adds `appliedRuleSet` in `GroupExpr` to avoid re-apply rules.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

### CODE REVIEW

File: planner/memo/group_expr.go [hunk 1]
```diff
@@ -32,14 +32,19 @@ type GroupExpr struct {
 	Group    *Group
 
 	selfFingerprint string
+	// appliedRuleSet saves transformation rules which have been applied to this
+	// GroupExpr, and will not be applied again. Use `interface{}` instead of
+	// `Transformation` to avoid import cycle.
+	appliedRuleSet map[interface{}]bool
```
Comment (lzmhhh123) [inline planner/memo/group_expr.go]: Can we use `map[interface{}]struct{}`?

Comment (francis0407) [reply planner/memo/group_expr.go]: ok, I have changed it.

File: planner/memo/group_expr.go [hunk 2]
```diff
@@ -63,3 +68,14 @@ func (e *GroupExpr) SetChildren(children ...*Group) {
 func (e *GroupExpr) Schema() *expression.Schema {
 	return e.Group.Prop.Schema
 }
+
+// AddAppliedRule adds a rule into the appliedRuleSet.
+func (e *GroupExpr) AddAppliedRule(rule interface{}) {
```
Comment (lamxTyler) [inline planner/memo/group_expr.go]: Not used in this PR?

Comment (winoros) [inline planner/memo/group_expr.go]: Where do we call this method?

Comment (francis0407) [reply planner/memo/group_expr.go]: Yes, but will be used in #13996.

Comment (francis0407) [reply planner/memo/group_expr.go]: Inside the transformation, like
```golang
// OnTransform implements Transformation interface.
func (rule *fakeTransformation) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
    Do the transformation...
    newExpr.AddAppliedRule(rule)
    return []*memo.GroupExpr{newExpr}, true, false, nil
}
```

**Comment by francis0407** [comment]
https://github.com/pingcap/tidb/pull/14078 also need this feature.

File: planner/cascades/optimize.go [hunk 3]
```diff
@@ -183,6 +183,10 @@ func (opt *Optimizer) findMoreEquiv(g *memo.Group, elem *list.Element) (eraseCur
 		// the transformation rule to enumerate all the possible expressions.
 		iter := memo.NewExprIterFromGroupElem(elem, pattern)
 		for ; iter != nil && iter.Matched(); iter.Next() {
+			if iter.GetExpr().HasAppliedRule(rule) {
```
Comment (eurekaka) [inline planner/cascades/optimize.go]: Check this in `Match`?

Comment (francis0407) [reply planner/cascades/optimize.go]: Agree. Then we only need this check for rules that can be added into the set.

**Pull request merged**